File: python/examples/time_travel.py
"""Time-travel: retrieve a law's full text at a historical git commit.

Usage::

    LEGALIZE_API_KEY=leg_... python examples/time_travel.py es ley_organica_3_2018
"""

from __future__ import annotations

import os
import sys

from legalize import Legalize


def main() -> int:
    if len(sys.argv) < 3:
        print("usage: time_travel.py <country> <law_id>", file=sys.stderr)
        return 2
    country, law_id = sys.argv[1], sys.argv[2]

    with Legalize(api_key=os.environ["LEGALIZE_API_KEY"]) as client:
        commits = client.laws.commits(country=country, law_id=law_id)
        print(f"{len(commits.commits)} commits for {law_id}\n")
        for _i, c in enumerate(commits.commits[:5]):
            print(f"  {c.sha[:10]}  {c.date}  {c.message.splitlines()[0][:60]}")

        if not commits.commits:
            return 0

        # Show the first historical version.
        oldest = commits.commits[-1]
        print(f"\n--- {law_id} at {oldest.sha[:10]} ({oldest.date}) ---\n")
        snapshot = client.laws.at_commit(country=country, law_id=law_id, sha=oldest.sha)
        print(snapshot.content_md[:600])
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: node/examples/time_travel.ts
/**
 * Time-travel: fetch a law at a past revision.
 *
 * Usage:
 *
 *   LEGALIZE_API_KEY=leg_... npx tsx examples/time_travel.ts es ley_organica_3_2018
 */

import { Legalize } from "legalize";

async function main(): Promise<number> {
  const country = process.argv[2] ?? "es";
  const lawId = process.argv[3] ?? "ley_organica_3_2018";

  const client = new Legalize();
  try {
    const commits = await client.laws.commits(country, lawId);
    console.log(`Found ${commits.commits.length} commits for ${lawId}.`);
    if (commits.commits.length === 0) return 0;
    const firstSha = commits.commits[commits.commits.length - 1]!.sha;
    console.log(`Retrieving law at first commit ${firstSha}...`);
    const past = await client.laws.atCommit(country, lawId, firstSha);
    console.log("---");
    console.log(past.content_md.slice(0, 500));
  } finally {
    await client.close();
  }
  return 0;
}

main().then((code) => process.exit(code));
